zebra-state/src/service.rs

```diff
@@ -252,6 +252,87 @@ impl StateService {
     pub fn utxo(&self, outpoint: &transparent::OutPoint) -> Option<transparent::Output> {
         self.mem.utxo(outpoint).or_else(|| self.sled.utxo(outpoint))
     }
+
+    /// Return an iterator over the relevant chain of the block identified by
+    /// `hash_or_height`.
+    #[allow(dead_code)]
+    pub fn chain(&self, hash_or_height: HashOrHeight) -> Iter<'_> {
+        Iter {
+            service: self,
+            state: State::NonFinalized(hash_or_height),
+        }
+    }
+}
+
+struct Iter<'a> {
+    service: &'a StateService,
+    state: State,
+}
+
+enum State {
+    NonFinalized(HashOrHeight),
+    Finalized(block::Height),
+    Finished,
+}
+
+impl<'a> Iter<'a> {
+    fn next_non_finalized_block(&mut self) -> Option<Arc<Block>> {
+        let Iter { service, state } = self;
+
+        let hash_or_height = match state {
+            State::NonFinalized(hash_or_height) => hash_or_height,
+            State::Finalized(_) | State::Finished => unreachable!(),
+        };
+
+        if let Some(block) = service.mem.block(*hash_or_height) {
+            let hash = block.header.previous_block_hash;
+            self.state = State::NonFinalized(hash.into());
```

Comment:
Can the caller hold the iterator across a chain reorganisation?

If the caller supplies a height on the best chain, this code turns the iterator into hash-based iterator on the best chain at the time the iterator's `next` method is first called.

Instead, let's convert to a hash when the iterator is created, so we choose the best chain at that time. I think those semantics are a bit clearer.

And let's document this potentially surprising behaviour in the function doc comment.

Reply:
> Can the caller hold the iterator across a chain reorganisation?

No, the caller won't be able to mutate the StateService at all while the iterator exists. Which should also mean that there's no reason to eagerly convert the height to a hash.

Reply:
Is this a standard property of rust iterators?
Or should we document it as part of the `chain` function docs?

Reply:
It's a property of borrows, this `Iter` type wraps a live shared borrow of the `StateService`, which means it would be a compiler error to produce a mutable borrow to that same service, or any of its member data.